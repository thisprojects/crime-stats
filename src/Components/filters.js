import React, { Component } from "react";

export const createListOfFilterCategories = categories =>
  categories.reduce((acc, curr) => {
    if (!acc[curr.category]) {
      acc[curr.category] = {};
      acc[curr.category].count = 1;
    } else {
      acc[curr.category].count++;
    }
    return acc;
  }, {});

export class CreateFilterItem extends Component {
  state = {
    checked: false
  };

  categoryRef = React.createRef();

  // If the filter checkbox state changes, 
  // pass the name of the filter to parent components add/remove filter method.  
  handleChecked = () => {
    this.setState(state => (state.checked = !state.checked));
    this.props.addAndRemoveFilters(this.categoryRef.current.id);
  };

  render() {
    const { category, count, totalCrimeCount, index } = this.props;
    return (
      <span
        key={ index }
        id={ category }
        ref={ this.categoryRef }
      >
        <label key={ index + 1 }>
          <input
            key={ index + 2 }
            type="checkbox"
            name={ category }
            defaultChecked={ this.state.checked }
            value={ category }
            onChange={ this.handleChecked }
          />
          <span key={ index + 3 } className="filter-item-wrapper">
            <p key={ index + 4 }>
              { category }
            </p>
            <p key={ index + 5 }>
              { count } Counts
            </p>
            <p key={ index + 6 }>
              { ((count / totalCrimeCount) * 100).toFixed(2) } %
            </p>
          </span>
        </label>
      </span>
    );
  }
}

class FilterCrimeResults extends Component {
  state = {
    currentlyAppliedFilters: []
  };

  remove = filterItem =>
    this.state.currentlyAppliedFilters.filter(item => item !== filterItem);

  add = filterItem => this.state.currentlyAppliedFilters.concat(filterItem);

  // Filter items are passed in from the CreateFilterItem component 
  addAndRemoveFilterItems = filterItem => {
    let currentlyAppliedFilters;
    
    if (this.state.currentlyAppliedFilters.includes(filterItem)) {
      currentlyAppliedFilters = this.remove(filterItem);
    } else {
      currentlyAppliedFilters = this.add(filterItem);
    }

    //Update local filter state and pass list of applied filters to parent 
    this.setState({ currentlyAppliedFilters }, () =>
      this.props.applyFilters(this.state.currentlyAppliedFilters)
    );
  };

  render() {
    const { totalCrimeCount, results } = this.props;
    const filterList = createListOfFilterCategories(results);

      return Object.keys(filterList).map((category, index) => (
            <CreateFilterItem
              key={ index }
              addAndRemoveFilters={ this.addAndRemoveFilterItems }
              index={ index }
              category={ category }
              count={ filterList[category].count }
              totalCrimeCount={ totalCrimeCount }
            />
          )
      );
  }
}

export default FilterCrimeResults;
