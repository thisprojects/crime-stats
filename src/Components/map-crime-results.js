import React from "react";

const ResultItems = ({ result }) => (
  <React.Fragment>
    <span className="result-item">{ result.category }</span>
    <span className="result-item">{ result.location.street.name }</span>
    <span className="result-item">
      { (result["outcome_status"] && result["outcome_status"].category) ||
        "No Outcome Recorded" }
    </span>
  </React.Fragment>
);

const MapCrimeResults = ({ results }) => {
  if (results.length > 0) {
    return results.map((result, index) => (
      <div key={ index } className="stat-row">
        <ResultItems result={ result } />
      </div>
    ));
  } else {
    return (
      <div className="no-results">
        <p>
          No Results - Click start to choose a location. Results are only
          available for UK locations
        </p>
      </div>
    );
  }
};

export default MapCrimeResults;
