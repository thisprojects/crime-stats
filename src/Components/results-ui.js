import React, { Component } from "react";
import FilterCrimeResults from "./filters";
import SortCrimeResults from "./sort";
import Slider from "./slider";
import MapCrimeResults from "./map-crime-results"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faFilter} from "@fortawesome/free-solid-svg-icons";



export class ResultsUI extends Component {
  state = {
    filteredAndSortedResults: this.props.defaultFilter,
    hideFilters: true
  };

  // Recieve active filters array from FilterCrimeResultss component and apply filters to results
  applyFilters = listOfActiveFilters => {
    this.setState(
      state =>
        (state.filteredAndSortedResults = this.props.results.filter(item =>
          listOfActiveFilters.includes(item.category)
        ))
    );
  };

  // Receive sortType from SortCrimeResults component and apply sort to results.
  sort = sortType => {
    if (this.state.filteredAndSortedResults.length > 0) {
      this.setState({
        filteredAndSortedResults: this.state.filteredAndSortedResults.sort(
          sortType
        )
      });
    } else {
      this.setState({
        filteredAndSortedResults: this.props.results.sort(sortType)
      });
    }
  };

  toggleFilterMenu = () => {
    this.setState(state => (state.hideFilters = !state.hideFilters));
  };

  render() {
    const { results } = this.props;
    const { filteredAndSortedResults } = this.state;
    
    return (
      <div className="crime-stats-wrapper">
        <div className="crime-stats">
          <SortCrimeResults totalCrimes={ results.length } sort={ this.sort } />
          <div className="results-table">
            <MapCrimeResults
              results={
                filteredAndSortedResults.length
                  ? filteredAndSortedResults
                  : results
              }
            />
          </div>
          <div className="filter-stub" onClick={ this.toggleFilterMenu }>
            <p>Filter</p>
            <FontAwesomeIcon icon={ faFilter } />
          </div>
          <Slider
            slide={ "top-slide" }
            hide={ "top-hide" }
            manualToggle={ this.state.hideFilters }
          >
            <div className="filter-wrapper">
              <FilterCrimeResults
                results={ results }
                totalCrimeCount={ results.length }
                applyFilters={ this.applyFilters }
              />
            </div>
          </Slider>
          ;
        </div>
      </div>
    );
  }
}

export default ResultsUI;
