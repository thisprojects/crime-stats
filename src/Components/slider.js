import React, { Component } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

export class Slider extends Component {
  sliderRef = React.createRef();

  toggleSlider = () => {
    this.sliderRef.current.classList.toggle(this.props.hide);
  };

  componentDidUpdate () {
    if (this.props.manualToggle === false) {
      this.sliderRef.current.classList.remove(this.props.hide)
    } else if (this.props.manualToggle === true) {
      this.sliderRef.current.classList.add(this.props.hide)
    }
  }

  render() {
    const {
      slide,
      hide,
      children,
      label,
      icon,
      stubClassName
    } = this.props;

    return (
      <div className={ `${slide} ${hide}` } ref={ this.sliderRef }>
        <div className={ `${stubClassName}` } onClick={ this.toggleSlider }>
          { icon && <FontAwesomeIcon icon={ icon } /> }
          <p>{ label }</p>
        </div>
        { children }
      </div>
    );
  }
}

export default Slider;
