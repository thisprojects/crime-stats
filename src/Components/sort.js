import React from 'react'

export const locationSort = (a, b) => {
  if (a.location.street.name < b.location.street.name) return -1;
  if (a.location.street.name > b.location.street.name) return 1;
  return 0;
};

export const outcomeSort = (a, b) => {
  if (a["outcome_status"] === null) return 1;
  if (b["outcome_status"] === null) return -1;
  if (a["outcome_status"].category < b["outcome_status"].category) return -1;
  if (a["outcome_status"].category > b["outcome_status"].category) return 1;
  return 0;
};


const SortCrimeResults = ({ sort, totalCrimes }) => 
  <div className="total-crimes">
    <p>Total Crimes { totalCrimes }</p>
    <input
      type="radio"
      name="sort"
      value="location"
      onChange={ () => sort(locationSort) }
    />
    <p>Location Sort</p>
    <input
      type="radio"
      name="sort"
      value="outcome"
      onChange={ () => sort(outcomeSort) }
    />
    <p>Outcome Sort</p>
  </div>

  export default SortCrimeResults