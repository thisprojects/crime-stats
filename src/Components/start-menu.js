import React, { Component } from "react";
import Slider from "./slider";
import { faPlayCircle } from "@fortawesome/free-solid-svg-icons";

export const generateYears = () => {
  let newDate = new Date();
  let currentYear = newDate.getFullYear();
  let years = [currentYear];
  for (let i = 0; i < 10; i++) {
    years.push(currentYear - (i + 1));
  }
  return years;
};

const months = [
  "01",
  "02",
  "03",
  "04",
  "05",
  "06",
  "07",
  "08",
  "09",
  "10",
  "11",
  "12"
];

export const Month = ({ handleChange }) => (
  <div className="date-month">
    <span>Month</span>
    <select id="month" onChange={ handleChange }>
      <GenerateOptions optionArray={ months } />
    </select>
  </div>
);

export const Year = ({ handleChange }) => (
  <div className="date-year">
    <span>Year</span>
    <select id="year" defaultValue={ "2019" } onChange={ handleChange }>
      <GenerateOptions optionArray={ generateYears() } />
    </select>
  </div>
);

export const GenerateOptions = ({ optionArray: array }) =>
  array.map((item, index) => (
    <option key={ index } value={ item }>
      { item }
    </option>
  ));

class StartMenu extends Component {
  state = {
    month: "01",
    year: "2019",
    hideStartMenu: false
  };

  locationInputText = React.createRef();

  handleSelectChange = ({ target: { value, id } }) => {
    this.setState((state) => (state[id] = value, state.hideStartMenu = false));
  };

  handleSubmit = () => {
    this.props.submit(
      `${this.locationInputText.current.value} uk`,
      this.state.year,
      this.state.month
    );
    this.setState({ hideStartMenu: true });
  };

  render() {
    return (
      <Slider
        slide={ "side-slide" }
        hide={ "side-hide" }
        label={ "Start" }
        icon={ faPlayCircle }
        manualToggle={ this.state.hideStartMenu }
        stubClassName={ "header-stub" }
      >
        <div className="start-menu-control-wrapper">
          <div className="start-menu-control">
            <div className="start-menu-inputs">
              <input
                id="post-code"
                type="text"
                placeholder="Location"
                ref={ this.locationInputText }
              />
              <Month handleChange={ this.handleSelectChange } />
              <Year handleChange={ this.handleSelectChange } />
            </div>
            <div className="submit-button">
              <button onClick={ this.handleSubmit }>Submit</button>
            </div>
          </div>
        </div>
      </Slider>
    );
  }
}

export default StartMenu
