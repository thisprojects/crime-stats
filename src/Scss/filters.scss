.filter-item-wrapper {
  display: flex;
  flex-direction: column;
  text-align:center;
  align-items: center;
  cursor: pointer;
  padding-bottom: 10px;
  border-bottom: 1px solid gray;
}

.filter-stub {
  width: 210px;
  padding: 7px 20px 7px 20px;
  text-align: center;
  position: absolute;
  top: 0;
  left: 60%;
  background-color: rgba(50, 50, 50, 0.9);
  z-index: 20;
  box-shadow: rgba(0, 0, 0, 0.3) 0px 1px 4px -1px;
  border-radius: 2px;
  cursor: pointer;
  color: white;
  svg {
    display: none;
  }
}

input[type="checkbox"] {
  display: none;
}

input[type="checkbox"] + p {
  color: #ccc;
}
input[type="checkbox"]:checked + .filter-item-wrapper p{
  color: white;
}