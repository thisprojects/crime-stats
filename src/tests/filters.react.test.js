import React from 'react';
import {createListOfFilterCategories, CreateFilterItem } from './../Components/filters'
import {
  mockResults, 
} from './mocked-results.js'
import renderer from 'react-test-renderer';


test('Filters Matches Snapshot', () => {
  const component = renderer.create(
    <CreateFilterItem
      key={ 1 }
      addAndRemoveFilters={ () => {} }
      index={ 1 }
      category={ 'fake-cat'}
      count={ 1 }
      totalCrimeCount={ '1337' }
    />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Create filter list function output as expected', () => {
  let x = createListOfFilterCategories(mockResults)
  console.log(x)
  expect(x).toEqual(   {
    'ccc Mock Category 1': { count: 1 },
    'bbb Mock Category 2': { count: 1 },
    'aaa Mock Category 3': { count: 1 }
  });
});