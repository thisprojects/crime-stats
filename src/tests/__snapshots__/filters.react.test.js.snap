// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Matches Snapshot 1`] = `
<span
  id="fake-cat"
>
  <label>
    <input
      defaultChecked={false}
      name="fake-cat"
      onChange={[Function]}
      type="checkbox"
      value="fake-cat"
    />
    <span
      className="filter-item-wrapper"
    >
      <p>
        fake-cat
      </p>
      <p>
        1
         Counts
      </p>
      <p>
        0.07
         %
      </p>
    </span>
  </label>
</span>
`;
